pydicom/filereader.py

```diff
@@ -282,6 +294,12 @@ def data_element_generator(fp,
         if length != 0xFFFFFFFF:
             # don't defer loading of Specific Character Set value as it is
             # needed immediately to get the character encoding for other tags
+            if tag_set is not None and tag not in tag_set and tag != (
```

Comment:
@massich, @darcymason: 
I updated the commit with a combination of your suggestions, please check if that is ok for you.
About the tests - I haven't updated them yet to cover tags with undefiend length.  Before starting to look around, do you know about a good test dataset for this in the test files?